ext/json/json_encoder.c

```diff
@@ -93,19 +93,26 @@ static inline void php_json_encode_double(smart_str *buf, double d, int options)
 }
 /* }}} */
 
-#define PHP_JSON_HASH_PROTECT_RECURSION(_tmp_ht) \
-	do { \
-		if (_tmp_ht) { \
-			GC_TRY_PROTECT_RECURSION(_tmp_ht); \
-		} \
-	} while (0)
-
-#define PHP_JSON_HASH_UNPROTECT_RECURSION(_tmp_ht) \
-	do { \
-		if (_tmp_ht) { \
-			GC_TRY_UNPROTECT_RECURSION(_tmp_ht); \
-		} \
-	} while (0)
+static inline zend_result php_json_protect_recursion(php_json_encoder *encoder, zend_refcounted *rc)
+{
+	if (GC_FLAGS(rc) & GC_IMMUTABLE) {
+		return SUCCESS;
+	}
+	if (zend_hash_index_add_empty_element(&encoder->recursive, (uintptr_t) rc)) {
```

Comment:
Leaving a note if anyone bases functionality on this in the future or restores this:

As I'd discovered in https://github.com/php/php-src/pull/7690 - using pointers as hash indexes directly leads to a lot of hash collisions and performance issues. Shifting by ZEND_MM_ALIGNED_OFFSET_LOG2 instead helps noticeably (to work with both malloc and emalloc)

(e.g. if data is aligned to 16 bytes (128 bits) in practice on a platform, then they'll all collide on the same 1 in 128 hash buckets)